File: solutions/swift/bottle-song/1/Sources/BottleSong/BottleSong.swift
struct BottleSong {
  private let numbers = [
    "Ten", "Nine", "Eight", "Seven", "Six", "Five", "Four", "Three", "Two", "One", "No",
  ]
  let bottles: Int

  func song(takedown: Int) -> [String] {
    let start = 10 - bottles
    return (start..<start + takedown).reduce(into: []) { (s, i) in
      if !s.isEmpty { s.append("") }
      let b1 = (i == numbers.count - 2) ? "bottle" : "bottles"
      let b2 = (i + 1 == numbers.count - 2) ? "bottle" : "bottles"
      s.append("\(numbers[i]) green \(b1) hanging on the wall,")
      s.append(s.last!)
      s.append("And if one green bottle should accidentally fall,")
      s.append("There'll be \(numbers[i + 1].lowercased()) green \(b2) hanging on the wall.")
    }
  }
}

File: solutions/ruby/bottle-song/1/bottle_song.rb
# frozen_string_literal: true

# module BottleSong
module BottleSong
  NUMBERS = %w[no one two three four five six seven eight nine ten].freeze

  def self.recite(start_bottles, take_down) =
    start_bottles
      .downto(start_bottles - take_down + 1)
      .map { verse(_1) }
      .join("\n")

  def self.verse(count) =
    <<~VERSE
      #{bottles(count).capitalize} hanging on the wall,
      #{bottles(count).capitalize} hanging on the wall,
      And if one green bottle should accidentally fall,
      There'll be #{bottles(count - 1)} hanging on the wall.
    VERSE

  def self.bottles(count) = "#{NUMBERS[count]} green bottle#{'s' unless count == 1}"
end
